File: remediation-runs/2026-09-06-approved-policy-completion/root/kotlin-container-attempt-01/ComposeContainerViewController.kt
@file:OptIn(kotlinx.cinterop.ExperimentalForeignApi::class)

package com.parlor.app

import platform.UIKit.NSLayoutConstraint
import platform.UIKit.UIInterfaceOrientation
import platform.UIKit.UIInterfaceOrientationMask
import platform.UIKit.UIViewController

/**
 * Gives SwiftUI its own outer view while Compose owns the child's semantics.
 * Host updates must not compete with the language provider's explicit native
 * direction on the Compose view. Navigation and session ownership stay in Compose.
 */
internal class ComposeContainerViewController(
    private val contentController: UIViewController,
) : UIViewController(nibName = null, bundle = null) {
    override fun viewDidLoad() {
        super.viewDidLoad()
        addChildViewController(contentController)
        val contentView = contentController.view
        contentView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(contentView)
        // Full physical bounds, not directional or safe-area anchors: parent
        // and child languages may differ, and Compose already handles insets.
        NSLayoutConstraint.activateConstraints(listOf(
            contentView.leftAnchor.constraintEqualToAnchor(view.leftAnchor),
            contentView.rightAnchor.constraintEqualToAnchor(view.rightAnchor),
            contentView.topAnchor.constraintEqualToAnchor(view.topAnchor),
            contentView.bottomAnchor.constraintEqualToAnchor(view.bottomAnchor),
        ))
        contentController.didMoveToParentViewController(this)
    }

    override fun childViewControllerForStatusBarStyle(): UIViewController = contentController

    override fun childViewControllerForStatusBarHidden(): UIViewController = contentController

    override fun childViewControllerForHomeIndicatorAutoHidden(): UIViewController = contentController

    override fun childViewControllerForScreenEdgesDeferringSystemGestures(): UIViewController = contentController

    override fun supportedInterfaceOrientations(): UIInterfaceOrientationMask =
        contentController.supportedInterfaceOrientations()

    override fun preferredInterfaceOrientationForPresentation(): UIInterfaceOrientation =
        contentController.preferredInterfaceOrientationForPresentation()

    override fun shouldAutorotate(): Boolean = contentController.shouldAutorotate()
}

File: iosApp/iosApp/ComposeContainerViewController.swift
import UIKit

/// Gives SwiftUI its own outer view while Compose owns the child's semantics.
/// Host updates must not compete with the language provider's explicit native
/// direction on the Compose view. Navigation and session ownership stay in Compose.
final class ComposeContainerViewController: UIViewController {
    private let contentController: UIViewController

    init(contentController: UIViewController) {
        self.contentController = contentController
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("Create this container with its content controller")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        addChild(contentController)
        let contentView = contentController.view!
        contentView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(contentView)
        // Full physical bounds, not directional or safe-area anchors: parent
        // and child languages may differ, and Compose already handles insets.
        NSLayoutConstraint.activate([
            contentView.leftAnchor.constraint(equalTo: view.leftAnchor),
            contentView.rightAnchor.constraint(equalTo: view.rightAnchor),
            contentView.topAnchor.constraint(equalTo: view.topAnchor),
            contentView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        ])
        contentController.didMove(toParent: self)
    }

    override var childForStatusBarStyle: UIViewController? { contentController }
    override var childForStatusBarHidden: UIViewController? { contentController }
    override var childForHomeIndicatorAutoHidden: UIViewController? { contentController }
    override var childForScreenEdgesDeferringSystemGestures: UIViewController? { contentController }

    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        contentController.supportedInterfaceOrientations
    }

    override var preferredInterfaceOrientationForPresentation: UIInterfaceOrientation {
        contentController.preferredInterfaceOrientationForPresentation
    }
}
